src/main/scala/com/advancedtelematic/campaigner/db/Campaigns.scala

```diff
@@ -107,21 +107,30 @@ protected [db] class Campaigns(implicit db: Database, ec: ExecutionContext)
   }
 
   /**
-    * Returns the IDs of all the devices that were processed in the campaign with `campaignId` and failed with
-    * the code `failureCode`.
+    * Returns the most recent device updates that have failed with the code `failureCode` in the campaign with ID
+    * `mainCampaignId` or any of its retry-campaigns.
     */
-  def fetchFailedDevices(campaignId: CampaignId, failureCode: String): Future[Set[DeviceId]] =
-    deviceUpdateRepo.findFailedByFailureCode(campaignId, failureCode)
+  private def findLatestFailedUpdates(mainCampaignId: CampaignId, failureCode: String): DBIO[Set[DeviceUpdate]] =
+    campaignRepo
+      .findRetryCampaignsOfAction(mainCampaignId)
+      .flatMap(cids => findFailedDeviceUpdatesAction(cids.map(_.id) + mainCampaignId))
+      .map(_.filter(_.resultCode.contains(failureCode)))
 
   /**
-    * Returns all the devices that are now failed with the code `failureCode` as a result of running the campaign
-    * with ID `campaignId` or any of its retry-campaigns.
+    * Returns the deviceIds of the most recent device updates that have failed with the code `failureCode`
+    * in the campaign with ID `mainCampaignId` or any of its retry-campaigns.
     */
-  def fetchFailureCodes(campaignId: CampaignId, failureCode: String): Future[Set[(DeviceId, String, String)]] = db.run {
-    campaignRepo
-      .findRetryCampaignsOfAction(campaignId)
-      .flatMap(cids => findFailedDeviceUpdatesAction(cids.map(_.id) + campaignId))
-      .map(_.filter(_.resultCode.contains(failureCode)))
+  def fetchLatestFailedDevices(campaignId: CampaignId, failureCode: String): Future[Set[DeviceId]] = db.run {
```

Comment:
I would suggest we create a generic `def fetchFailedDevices(campaignId: CampaignId, failureCode: String): Future[Set[DeviceUpdate]]` in `Campaigns` class and left it to caller to extract attributes it needs. Each of these two method has exactly one usage and I think the knowledge of what exactly caller needs should be a part of `Campaigns`

Reply:
Makes sense. Updated.